infrastructure/network/addressmanager/addressmanager.go

```diff
@@ -5,1264 +5,254 @@
-	if err != nil {
-		return nil, err
-	}
-
-	return am.HostToNetAddress(host, uint16(port), appmessage.SFNodeNetwork)
-}
-
-// Start begins the core address handler which manages a pool of known
-// addresses, timeouts, and interval based writes.
-func (am *AddressManager) Start() error {
-	// Already started?
-	if atomic.AddInt32(&am.started, 1) != 1 {
-		return nil
-	}
-
-	log.Trace("Starting address manager")
-
-	// Load peers we already know about from the database.
-	err := am.loadPeers()
-	if err != nil {
-		return err
-	}
-
-	// Start the address ticker to save addresses periodically.
-	am.wg.Add(1)
-	spawn("addressManager.addressHandler", am.addressHandler)
-	return nil
 }
 
-// Stop gracefully shuts down the address manager by stopping the main handler.
-func (am *AddressManager) Stop() error {
-	if atomic.AddInt32(&am.shutdown, 1) != 1 {
-		log.Warnf("Address manager is already in the process of " +
-			"shutting down")
-		return nil
-	}
-
-	log.Infof("Address manager shutting down")
-	close(am.quit)
-	am.wg.Wait()
-	return nil
-}
-
-// AddAddresses adds new addresses to the address manager. It enforces a max
-// number of addresses and silently ignores duplicate addresses. It is
-// safe for concurrent access.
-func (am *AddressManager) AddAddresses(addresses []*appmessage.NetAddress, sourceAddress *appmessage.NetAddress, subnetworkID *subnetworkid.SubnetworkID) {
+// Addresses returns all addresses
+func (am *AddressManager) Addresses() []*appmessage.NetAddress {
 	am.mutex.Lock()
 	defer am.mutex.Unlock()
 
-	for _, address := range addresses {
-		am.updateAddress(address, sourceAddress, subnetworkID)
+	result := make([]*appmessage.NetAddress, 0, len(am.addresses))
+	for _, address := range am.addresses {
+		result = append(result, address.netAddress)
 	}
+	return result
 }
 
-// AddAddress adds a new address to the address manager. It enforces a max
-// number of addresses and silently ignores duplicate addresses. It is
-// safe for concurrent access.
-func (am *AddressManager) AddAddress(address, sourceAddress *appmessage.NetAddress, subnetworkID *subnetworkid.SubnetworkID) {
+// NotBannedAddressesWithException returns all not banned addresses with excpetion
+func (am *AddressManager) NotBannedAddressesWithException(exceptions []*appmessage.NetAddress) []*appmessage.NetAddress {
+	exceptionsKeys := netAddressesKeys(exceptions)
 	am.mutex.Lock()
 	defer am.mutex.Unlock()
 
-	am.updateAddress(address, sourceAddress, subnetworkID)
-}
-
-// numAddresses returns the number of addresses that belongs to a specific subnetwork id
-// which are known to the address manager.
-func (am *AddressManager) numAddresses(subnetworkID *subnetworkid.SubnetworkID) int {
-	if subnetworkID == nil {
-		return am.fullNodeNewAddressCount + am.fullNodeTriedAddressCount
+	result := make([]*appmessage.NetAddress, 0, len(am.addresses))
+	for key, address := range am.addresses {
+		if !address.isBanned && !exceptionsKeys[key] {
```

Comment:
It might make sense to keep all the banned addresses in a separate list as well, instead of filtering them out every time we need an address